app/features/worksheet_generator/tools.py

```diff
@@ -0,0 +1,481 @@
+        if topic is None: raise ValueError("Topic must be provided")
+        if difficulty_level is None: raise ValueError("Difficulty level must be provided")
+        if question_types is None: raise ValueError("At lease one question type must be provided")
+
+    def create_worksheet(self):
+        params = {"grade_level": self.grade_level, "topic": self.topic, "difficulty_level": self.difficulty_level, "lang": self.lang, "verbose": self.verbose}
+
+        worksheet = []
+
+        for question_type in self.question_types:
+            if self.verbose: logger.info(f"Generating {question_type['question_type']} questions")
+
+            #generating the question list
+            question_list = create_question_builder(question_type['question_type'], **params).create_questions(question_type['num_questions'])
+            worksheet.append(question_list)
+
+        # Return the generated worksheet(list of lists that include questions of each type requested)
+        return worksheet
+
+# Abstract class for the question building functionality
+class QuestionBuilder(ABC):
+    @abstractmethod
+    def __init__(self):
+        pass
+
+    def generate_question(self):
+        # Return the generated response
+        prompt = PromptTemplate(
+            template=self.prompt,
+            input_variables=["grade_level", "topic", "difficulty_level", "lang"],
+            partial_variables={"format_instructions": self.parser.get_format_instructions()}
+        )
+
+        # Generate the final prompt by filling in the variables
+        filled_prompt = prompt.format(
+            grade_level=self.grade_level,
+            topic=self.topic,
+            difficulty_level=self.difficulty_level,
+            lang=self.lang
+        )
+
+        if self.verbose:
+            logger.info(f"Generated prompt: {filled_prompt}")
+
+        # Generate the question using the Google Gemini model
+        response = self.model.generate(prompts = [filled_prompt])
+
+        if self.verbose: logger.info("Response generation complete")
+    
+        # Extract the generated text from the response
+        output = response.generations[0][0].text.strip()
+
+        # Parse the JSON response using the configured parser
+        parsed_output = self.parser.parse(output)
+
+        if self.verbose: logger.info(f"Response parsing complete using {self.parser}")
+
+        return parsed_output
+    
+    @abstractmethod
+    def validate_response(self, response: Dict):
+        pass
+
+    @abstractmethod
+    def transform_json_dict(input_data: Dict) -> Dict:
+        pass
+
+    @abstractmethod
+    def create_questions(self, num_questions: int) -> List[Dict]:
+        pass
+
+class MultiChoiceQuestionBuilder(QuestionBuilder):
+    def __init__(self, grade_level, topic, difficulty_level, lang, prompt=None, model=None, parser=None, verbose=False):
+        default_config = {
+            "model": GoogleGenerativeAI(model="gemini-1.0-pro"),
```

Comment:
Refactor this model to `gemini-1.5-flash` or `gemini-1.5-pro` as v. 1.0 is not working anymore.